### ISSUE TITLE
Drag a copy

### ISSUE THREAD

**Issue by rosoft2001** [mcFrax/omicron-dnd#9]
Seems a very nice plugin, but I need your help to use it.

What would be the configuration to drag a copy from list1 into list2?

Thank you

**Comment by jan-warchol** [comment]
Hello! Nice to see that people are discovering this library :-)
Just to clarify: do you want it to work in one direction (copying from list1 to list2), or in both directions (from list1 to list2 and from list2 to list1)?

**Comment by mcFrax** [comment]
It means that you'd like to keep the element visible in list1, right?

That is currently not supported. It shouldn't be that hard to write, I think - it would basically mean skipping the special treatment of the source container and not much more - but it may be quite some work to avoid it getting messy.

At the moment I think the best thing you could do would be to add a handler `onDropToOtherContainer`, create a copy of the dragged element, and put it back in the old place (or, to keep the specific instance of element in place - revert the move and insert a copy into the new place).

Unfortunately, for a solution working out-of-the box the best thing I can do at the momement is to redirect you to [Sortable](https://github.com/SortableJS/Sortable), which is more feature-rich and has cloning implemented (I created Omicron as a more performant replacement of Sortable - the API is pretty similar).

I'll keep this in mind, this feature would be nice to have - I think I can implement it in a few weeks, if all goes well.

**Comment by rosoft2001** [comment]
Thanks for info, I'll keep an eye on new features. 

Speaking of the performance a benchmark would be also nice (from the samples is hard to tell).

**Comment by mcFrax** [comment]
Right, perhaps side-by-side comparison would show the difference well. I guess I can try to measure the FPS too.

The "performance" here is mainly about the responsiveness - Sortable gets very slow when DOM complexity increases.

**Comment by rosoft2001** [comment]
Performance is more than important. 

For my future project I need to copy from source into destination, not other way around.

In Sortable is: {pull: 'clone', put: false}

It may be a way to use onBeforeDragStart to make a copy and onContainerEntered to not allow drop?

**Comment by mcFrax** [comment]
I will be away for the weekend, so probably won't respond before Monday. Next week I plan to finally publish Omicron in npm, so it should be available that way, too - and if I have some spare cycles, I'll look into the cloning.

Hope that the solution with onBeforeDragStart works for you - good luck! 🙂

**Comment by mcFrax** [comment]
Uh, I just realized that `forbiddenInsertionIndicesFn: ()=>[]` is wrong, it actually needs to return the whole list of indices.

```
forbiddenInsertionIndicesFn(containerEl) {
    return [...Array.prototype.keys(containerEl.children), containerEl.children.length];
}
```

Should work. It returns all indices in containerEl.children and the last but one index (which may also be suitable for insertion).

Now, I think it would be neat to have some special value there, like `'ALL'`. Or simply add some counterpart to Sortable's `{put: false}`, as a simpler alternative to `forbiddenInsertionIndicesFn`.

I [published the library in npm](https://www.npmjs.com/package/omicron-dnd). I will aim to release any fixes quickly there too, so it should be a reliable source from now on.
